dask/array/reductions.py

```diff
@@ -294,7 +294,7 @@ def partial_reduce(
     dsk = {}
     for k, p in zip(keys, product(*parts)):
         decided = {i: j[0] for (i, j) in enumerate(p) if len(j) == 1}
-        dummy = dict(i for i in enumerate(p) if i[0] not in decided)
+        dummy = dict(i for i in enumerate(p) if i[0] in split_every)
```

Comment:
This isn't quite right (but I'm not sure what it should be). The keys in `decided` and `dummy` should never intersect (but `assert set(decided).union(dummy) == set(range(ndim))` should always be true). I suspect you probably want to change how `decided` is created, and leave the creation of `dummy` as it was before.

Reply:
I agree with your latter assertion:
```python
assert set(decided).union(dummy) == set(range(ndim))
```
but I differ on your claim that "The keys in `decided` and `dummy` should never intersect".

The following was my reasoning (pardon me if I'm needlessly going over what you already know):

* I discovered that `lol_tuples()` (on the next line after `dummy`) outputs a single array-like list, or list-of-lists (lol), `g` whose number of dimensions always equals the number of keys in `dummy`.  
   https://github.com/dask/dask/blob/6824e5dd7807971270d5828a61618ccd81592199/dask/array/reductions.py#L294-L299


* This lol `g` is then passed on to `_concatenate2()` together with the previously passed `axis` tuple (from `_tree_reduce()`) which never changes throughout the recursion.  https://github.com/dask/dask/blob/6824e5dd7807971270d5828a61618ccd81592199/dask/array/reductions.py#L232-L233

* Now since `_concatenate2()` needs a corresponding lol-dimension for each value of `axis`, then the length of `dummy` also needs to remain fixed even if its individual items change with each recursion.

* The keys of `decided`, on the other hand are all those axes whose number of chunks is equal to 1, either because the axis in question has been fully reduced, or because it is not taking part in the reduction, as can be deduced from following code-extract: https://github.com/dask/dask/blob/6824e5dd7807971270d5828a61618ccd81592199/dask/array/reductions.py#L280-L283

To confirm this, the following code snippet:
```python
from tlz import partition_all


split_every = {0: 2, 1: 2}	# keys here are the reduction axes
parts = [
    list(partition_all(split_every.get(i, 1), range(n)))
    for (i, n) in enumerate(da_x.numblocks)
]
```
yields:
```python
# when da_x.numblocks = (16, 8, 4)
parts = 
[[(0, 1), (2, 3), (4, 5), (6, 7), (8, 9), (10, 11), (12, 13), (14, 15)],
 [(0, 1), (2, 3), (4, 5), (6, 7)],
 [(0,), (1,), (2,), (3,)]]
```

* So I concluded that while `dummy` ought to remain fixed, `decided` will increase in length starting from the number of axes not taking part in the reduction.

Reply:
> but I differ on your claim that "The keys in decided and dummy should never intersect".

Sorry, I think I wasn't clear here. I'm not saying that how you're handling generating `dummy` is wrong, all I'm saying is that the following assertion should never trigger:

```
assert not set(dummy).intersection(decided)
```

Otherwise this code is relying on an implementation detail of `lol_tuples`, which checks `dummies` first before checking `decided`. If this implemention were to change at all then the existing code would be wrong. However you want to handle this is up to you, but that assertion should pass (note, I don't think you should add that assertion to the code, just that you should think about it in your code above).

Reply:
Oh I see.  You're right about that.  Sorry for the misunderstanding.  Hmm ...  I'll need to think about this some more.

Reply:
If you think your code is correct as is, then it might be as simple as just removing the intersecting ones from `dummy`. Not elegant, but functional. No strong thoughts though.

Reply:
Thanks @jcrist .  This was a good idea, and quite important too on deeper reflection.   I'm guessing that you meant to say, "remove the intersecting ones from `decided`," instead of `dummy`, right? 

I also renamed `decided` to `free`, which appears to be the designated antonym for `dummy` in the context of indices.  I hope it improves readability of the code. 

Well then, I suppose my latest commit completes this PR's review.  Would you agree?